[3기-C 복신영] JPA 게시판 구현 미션 제출합니다

Sinyoung3016 opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
JPA를 사용하여 간단한 게시판 API 만들기

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
- User와 Post 엔티티 구현
- Post API 구현
- 통합 및 단위 테스트
- rest docs를 사용하여 문서화

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
- 특정 api에서 발생할 수 있는 에러 또한 문서화를 해야하는 지 궁금합니다

Sinyoung3016 committed:
chore: 스프링 프로젝트 생성

Sinyoung3016 committed:
chore: 의존성 및 설정 정리

Sinyoung3016 committed:
feat: user와 저장소 생성

- user도 예약어인지 오류 발생하여, 테이블명 수정
- @CreatedDate 사용을 위해, @EnableJpaAuditing 추가

Sinyoung3016 committed:
feat: post와 저장소 생성

Sinyoung3016 committed:
test: Post 생성 및 조회 테스트

Sinyoung3016 committed:
feat: Post 서비스 생성 및 테스트

Sinyoung3016 committed:
feat: @ControllerAdvice를 사용하여 예외 처리

Sinyoung3016 committed:
feat: Post 컨트롤러 생성 및 통합 테스트

Sinyoung3016 committed:
chore: restdoc 의존성 설정

Sinyoung3016 committed:
docs: Post api 문서화

SeokRae commented on src/main/java/kdt/springbootboardjpa/service/PostService.java:
```diff
@@ -0,0 +1,52 @@
+
+@Service
+@AllArgsConstructor
+public class PostService {
```
Query & Command에 용도에 맞는 Transactionl 처리가 필요할 것 같습니다.!

Sinyoung3016 replied:
command와 query가 같이 존재하는 서비스 메소드의 경우, transaction을 적용해주면 될까요?

+기본적으로 @Transactional(readOnly=true)를 클래스에 추가하고, command가 같이 존재하는 경우 메소드에 @transactional을 따로 추가해보았습니다!

SeokRae reviewed (commented):
고생하셨습니다. 신영님 !

외부로 상태 값에 대한 의미를 노출하여야 하는 경우에는 RestDocs에 정의되는 것이 좋을 것 같습니다.
추가로 공통 응답 객체 같은 걸 만드시는 것을 추천드립니다. !

시큐리티도 화이팅 하시죠.!!

zxcv9203 commented on src/main/java/kdt/springbootboardjpa/controller/PostController.java:
```diff
@@ -0,0 +1,59 @@
+                .build();
+        Post savedPost = postService.createPost(newPostDto);
+        PostResponse response = PostResponse.toPostResponse(savedPost);
+        return new ResponseEntity<>(response, HttpStatus.CREATED);
```
CREATED라는 HTTP 상태코드를 통해 자원이 생성되었음을 명시해주는 것은 좋은 것 같습니다 👍 

 한발 더 나아가서 CREATED 헤더의 요구사항도 충족하는 것도 좋을 것 같습니다.
https://developer.mozilla.org/ko/docs/Web/HTTP/Status/201

Sinyoung3016 replied:
용철님, 궁금한 점이 있습니다!
말씀해주신 요구사항 충족을 위해서, created 응답 헤더에 Content-Location 값을 넣어주고 싶었습니다. 
근데 HttpHeaders에는 여러 헤더값에 대한 set메소드가 있던데, setContentLocation()은 없더라구요. 그래서 결국 직접 넣어줬습니다.
해당 메소드는 왜 없는 지 궁금하네요..

zxcv9203 replied:
직접 넣어주셨다는게 다음과 같은 형식을 말씀하시는 것이 맞을까요? 🤔 

``` java
HttpHeaders headers = new HttpHeaders();

headers.add(HttpHeaders.CONTENT_LOCATION, value);
```

그리고 201 응답에 대한 헤더의 이름은 Location이 맞습니다..! Location 헤더 같은 경우에는 set 메서드가 존재합니다.
`setContentLocation()` 메서드가 존재하지 않는 이유에 대해서는 저도 잘 모르겠네요 😅 

그리고 HttpHeader를 직접 넣어주시는 것도 좋지만 ResponseEntity의 빌더를 이용하면 다음과 같이 필수 헤더를 쉽게 적용할 수 있다는 생각이 드네요..!

```java
return ResponseEntity.created(URI)
    .body(response)
    .build();
```

Sinyoung3016 replied:
헐 진짜 오늘 왜이러지,, 둘을 반대로 생각하고 있었습니다!
바로 수정하겠습니다
+location의 경우에는 메소드가 존재하네요!
빌더도 참고하겠습니다~ 감사합니다!

zxcv9203 commented on src/main/java/kdt/springbootboardjpa/controller/request/SavePostRequest.java:
```diff
@@ -0,0 +1,20 @@
+
+@NoArgsConstructor
+@Getter
+public class SavePostRequest {
```
record 타입을 많이 사용하셨던 것으로 기억하는 record 타입이 아닌 class로 Request 객체를 생성하신 이유가 있으신가요? 🤔

Sinyoung3016 replied:
Builder를 사용하고 싶은데, record는 builder가 적용되지 않더라구요. 그래서 이번에는 class로 만들었습니다. 
혹시 용철님은 어떤 방식을 더 선호하시나요?

zxcv9203 replied:
저는 가능하면 record 타입으로 많이 사용하기는 합니다. 해당 부분은 인텔리제이 버그로 보입니당.. 

저는 실행은 잘되네용 🤔 

https://stackoverflow.com/questions/69825016/lomboks-builder-not-detecting-fields-of-the-java-record

```suggestion
public record SavePostRequest(
    String title,
    String content,
    Long createdBy
) {

    @Builder
    public SavePostRequest {
    }
}
```

위와 같이 사용하면 인텔리제이에서도 자동완성 기능 및 에러가 발생하지 않네요..! 😄

Sinyoung3016 replied:
와우 바로 적용했습니다! 더 자세히 알아볼껄 그랬네요ㅎㅎ 
감사합니다~

zxcv9203 commented on src/main/java/kdt/springbootboardjpa/controller/response/PostResponse.java:
```diff
@@ -0,0 +1,32 @@
+import lombok.NoArgsConstructor;
+
+@Getter
+@NoArgsConstructor
```
기본생성자를 추가해주신 이유가 궁금합니다. 🤔

zxcv9203 commented on src/main/java/kdt/springbootboardjpa/exception/ExceptionResolver.java:
```diff
@@ -0,0 +1,19 @@
+@ControllerAdvice
+public class ExceptionResolver {
+    @ExceptionHandler({IllegalArgumentException.class})
+    @ResponseStatus(HttpStatus.NOT_FOUND)
```
현재 ResponseEntity를 이용해서 상태 코드를 넣어주고 있어 @ResponseStatus는 지워주셔도 될 것 같습니다 😄

zxcv9203 commented on src/main/java/kdt/springbootboardjpa/respository/entity/Post.java:
```diff
@@ -0,0 +1,57 @@
+    @JoinColumn(name = "user_id")
+    private User user;
+
+    @CreatedDate
+    private LocalDateTime createdAt;
```
엔티티들에게 공통적으로 관리되는 것이라면 추상 클래스로 만들어 관리하시는 것은 어떠신가요? 😄

zxcv9203 commented on src/main/java/kdt/springbootboardjpa/respository/entity/User.java:
```diff
@@ -0,0 +1,57 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NotNull
+    @NotBlank
```
@NotBlank가 NULL 여부도 검증합니다..!

다음은 javadoc의 @NotBlank에 대한 설명입니다.

<img width="649" alt="image" src="https://user-images.githubusercontent.com/41960243/208455336-b93d1bc7-88b5-4e89-8c60-97aaf7fbe033.png">

zxcv9203 commented on src/main/java/kdt/springbootboardjpa/service/PostService.java:
```diff
@@ -0,0 +1,52 @@
+                .content(postDto.getContent())
+                .user(savedPost.getUser())
+                .build();
+        return postRepository.save(newPost);
```
Dirty Checking에 대해서 아시나요?! Update임에도 save를 해주신 이유가 궁금합니다. 🤔

zxcv9203 commented on src/main/java/kdt/springbootboardjpa/service/dto/PostDto.java:
```diff
@@ -0,0 +1,22 @@
+import lombok.NoArgsConstructor;
+
+@Getter
+@NoArgsConstructor
```
기본생성자를 넣어주신 이유가 궁금합니다. 🤔

zxcv9203 commented on src/test/java/kdt/springbootboardjpa/service/PostServiceTest.java:
```diff
@@ -0,0 +1,155 @@
+
+        assertThatThrownBy(
+                () -> postService.createPost(postDto)
+        ).isInstanceOf(RuntimeException.class)
```
예외는 발생시키는 예외가 같은 종류로 해야 원하는 예외가 발생했는지 확인할 수 있을 것 같습니다..! 😄

zxcv9203 reviewed (commented):
신영님 JPA Board 과제 고생하셨습니다 🎉🎉🎉🎉

1등 제출..! 축하드립니다!! 🥳 
 
엔티티에서 검증해주시는 것도 좋지만 처음 들어오는 시점에서 DTO에 bean validation을 추가해주셔도 좋을 것 같다는 생각이 듭니다..!

> 특정 api에서 발생할 수 있는 에러 또한 문서화를 해야하는 지 궁금합니다

특정 API에서 예상되는 응답이 있다면 (잘못된 인자가 들어왔을때 400, 자원을 찾을 수 없을때 404 등) 해당하는 에러에 대해 응답 형태를 확인해 볼 수 있으면 좋을 것 같습니다 👍

Sinyoung3016 committed:
refactor: 사용하지 않는 코드 지우기

- @NotBlank는 @NotNull을 포함하고 있음
- 사용하지 않는 기본 생성자 삭제

Sinyoung3016 committed:
refactor: 사용하지 않는 코드 지우기

- ResponseEntity를 통해 이미 만들어주고 있음

Sinyoung3016 committed:
refactor: 엔티티의 공통 부분을 묶어주는 추상클래스 생성

Sinyoung3016 committed:
test: 예외 클래스를 더 명확하게 검사

Sinyoung3016 committed:
refactor: 필요 없는 코드 지우기

- dirtycheck를 잊지말라구

Sinyoung3016 committed:
feat: CREATED 상태 응답 시에 header에 content-location 넣어주기

Sinyoung3016 committed:
refactor: dto를 class에서 record로 수정

Sinyoung3016 committed:
test: 사용하지 않은 코드 지우기

Sinyoung3016 committed:
docs: 에러 응답 문서화

Sinyoung3016 committed:
fix: CREATED 상태 응답 시 header에 location 넣어주기

Sinyoung3016 committed:
fix: Post Update 로직 수정

Sinyoung3016 committed:
refactor: Post 서비스에 transactional 추가

- Query & Command에 용도에 맞는 Transactionl 처리
- readOnly를 통해 dirtycheck를 유지하지 않아도 되 성능 향상
https://willseungh0.tistory.com/75

Sinyoung3016 committed:
refactor: 공통 응답 객체 생성

Sinyoung3016 committed:
refactor:  응답에서 중복되는 요소 삭제

Sinyoung3016 committed:
feat: PostDto에 @Valid 추가

SeokRae commented on src/main/java/kdt/springbootboardjpa/controller/request/SavePostRequest.java:
```diff
@@ -0,0 +1,11 @@
+import jakarta.validation.constraints.Positive;
+import lombok.Builder;
+
+public record SavePostRequest(@NotBlank String title, @NotBlank String content, @Positive Long createdBy) {
```
기본 생성자가 필요해서 record에 default constructor를 명시한 뒤에 @Builder를 마킹한 이유가 궁금하네요.!

그리고 constructor에 선언된 필드 들은 아래 IDE를 사용하여 가독성 있게 나열할 수 있습니다.

<img width="826" alt="CleanShot 2023-01-02 at 10 29 46@2x" src="https://user-images.githubusercontent.com/17922700/210189729-4f83f423-6535-4586-a7ed-59cfb282e508.png">

SeokRae commented on src/main/java/kdt/springbootboardjpa/controller/response/PostResponse.java:
```diff
@@ -0,0 +1,22 @@
+import kdt.springbootboardjpa.respository.entity.Post;
+import lombok.Builder;
+
+public record PostResponse(@Positive Long id, @NotBlank String title, @NotBlank String content, @Positive Long createdBy) {
```
이 부분도 SavePostRequest 커멘트와 동일합니다.

<img width="391" alt="CleanShot 2023-01-02 at 09 36 32@2x" src="https://user-images.githubusercontent.com/17922700/210188543-f641acdb-ed09-42fa-a06c-6978285f98e8.png">

SeokRae commented on src/main/java/kdt/springbootboardjpa/controller/request/SavePostRequest.java:
```diff
@@ -0,0 +1,11 @@
+import lombok.Builder;
+
+public record SavePostRequest(@NotBlank String title, @NotBlank String content, @Positive Long createdBy) {
+    @Builder
```
jackson 라이브러리의 ObjectMapper의 역직렬화(Deserialization)를 위해서 기본 생성자를 생성하셨다면
한번 없애고 시도 해보셨으면 좋겠네요.!

그리고 jackson 라이브러리는 
직렬화(object -> json)는 getter를 필요로 하고, 역 직렬화(json -> object)는 기본생성자를 필요로 합니다.

근데 record는 기본 생성자 없이도 직렬화 자연스럽게 되는 것 같은데 왜그런걸까요?

<img width="666" alt="CleanShot 2023-01-02 at 10 34 51@2x" src="https://user-images.githubusercontent.com/17922700/210189858-03e1fdef-78a4-4af9-b06a-427bc009472c.png">

SeokRae commented on src/main/java/kdt/springbootboardjpa/global/exception/ExceptionResolver.java:
```diff
@@ -0,0 +1,20 @@
+
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler({IllegalArgumentException.class})
+    public BaseResponse handleForbiddenException(HttpServletRequest request, RuntimeException e) {
```
```suggestion
    public BaseResponse handleForbiddenException(HttpServletRequest request, RuntimeException e) {
```

컴파일러에게 타입 추론을 맡겨야 하는 상황이 아니면 제네릭 타입을 명시해주는 것이 더 좋습니다.
더 좋은 이유는 컴파일러가 제네릭 Raw 타입을 어떻게 타입 추론하는지에 대해서 알아보면 알 수 있긴합니다만,
여기서는 일단 제네릭 타입을 명시하는 것이 좋다 정도만 알고 넘어가면 좋지 않을까 싶네요.!

SeokRae commented on src/main/java/kdt/springbootboardjpa/global/exception/ExceptionResolver.java:
```diff
@@ -0,0 +1,20 @@
+    @ExceptionHandler({IllegalArgumentException.class})
+    public BaseResponse handleForbiddenException(HttpServletRequest request, RuntimeException e) {
+        ApiError apiError = new ApiError(e.getMessage(), request.getRequestURI());
+        return BaseResponse.error(apiError);
```
IllegalArgumentException과 같은 오류는 사용자의 입력에 대한 오류로 볼 수 있을 것 같습니다.
사용자에게 잘 못 입력하였다는 메시지를 전달하는 것도 필요하지만, 서버에서 해당 로그를 확인할 수 있도록 로깅하는 것도 필수 일 것 같습니다.
log도 한번 찍어서 로그 파일에 남길 수 있도록 해보시죠 !

SeokRae reviewed (commented):
안녕하세요. 신영님 늦은 리뷰에 반성합니다. ㅎㅎ..

1차적으로 몇가지 커멘트 남겼습니다.

현재 프로젝트를 잘 다듬어 템픞릿처럼 사용할 수 있도록 잘 정리해보시죠.!!

화이팅 입니다.!

SeokRae commented on src/main/java/kdt/springbootboardjpa/respository/entity/Post.java:
```diff
@@ -0,0 +1,53 @@
+        this.user = user;
+    }
+
+    public void setUser(User user) {
```
딴건 update로 시작하는데 요 친구는 set으로 시작하는 군요.!
어떤 이유가 있을까요?

SeokRae commented on src/main/java/kdt/springbootboardjpa/service/PostService.java:
```diff
@@ -0,0 +1,53 @@
+    public Post updatePost(Long postId, @Valid PostDto postDto) {
+        Post savedPost = postRepository.findById(postId)
+                .orElseThrow(() -> new IllegalArgumentException("Can't find Post"));
+        savedPost.updateTitle(postDto.title());
+        savedPost.updateContent(postDto.content());
```
```suggestion
        savedPost.updateTitle(postDto.title());
        savedPost.updateContent(postDto.content());
```

두 필드에 대한 수정이 모두 일어난다면 update 메서드 하나만 만들고
 아예 엔티티 안에서 수정이 일어나도 좋을 것 같습니다.

SeokRae commented on src/main/java/kdt/springbootboardjpa/service/dto/PostDto.java:
```diff
@@ -0,0 +1,13 @@
+import jakarta.validation.constraints.PositiveOrZero;
+import lombok.Builder;
+
+public record PostDto(Long id, @NotBlank(message = "NotBlank") String title, @NotBlank(message = "NotBlank") String content, @Positive(message = "Positive") Long createdBy) {
```
```suggestion
public record PostDto(
	Long id,
	String title,
	String content,
	Long createdBy
) {
	//...
}
```

파라미터가 많아져서 길어진다면 아래 기능을 활용해주세요!

<img width="651" alt="CleanShot 2023-01-03 at 11 02 19@2x" src="https://user-images.githubusercontent.com/17922700/210291366-ae530132-7656-465c-a40e-21d70d3ec743.png">

SeokRae commented on src/test/java/kdt/springbootboardjpa/controller/PostControllerTest.java:
```diff
@@ -0,0 +1,262 @@
+@SpringBootTest
+@AutoConfigureMockMvc
+@AutoConfigureRestDocs
+public class PostControllerTest {
```
RestDocs 좋습니다.!!
